### ISSUE TITLE
plugin/ochttp: doesn't implement http.Hijacker

### ISSUE THREAD

**Issue by t-bast** [census-instrumentation/opencensus-go#642]
I'm using gorilla websockets on a /websockets route exposed by my http server.
I started using ochttp.Handler to get automagic instrumentation (which is great!) but it apparently doesn't allow hijacking the connection: gorilla errors out saying that the response doesn't implement ([http.Hijacker](https://golang.org/pkg/net/http/#Hijacker).

Do you think it's something that would make sense to implement in ochttp? Or do you recommend exposing my websocket endpoint with a normal http handler separately?

Thanks!

**Comment by ramonza** [comment]
We should consider implementing all the optional interfaces ResponsesWriter might implement in net/http: Flusher, Pusher, Hijacker, etc.

**Comment by odeke-em** [comment]
@t-bast is this with an HTTP/1 connection? Does your service allow http.ServeMux successfully when integrated with gorilla/mux?

@ramonza we might need to be conservative with the interfaces that we implement because http.Hijack and HTTP/2 connections are incompatible.

**Comment by odeke-em** [comment]
@t-bast no worries, I've been able to repro this and perhaps it'll serve as our non-regression integration test later on at https://gist.github.com/odeke-em/3c2c664416933f2b349cd662b0b63c68.

Actually someone, @basvanbeek might be interested in a gorilla/mux solution over the next couple of weeks when free.

**Comment by odeke-em** [comment]
Alright, after some investigation I've concluded that we can safely just make plugin/ochttp's response proxy the call to Hijacker because even the standard library on HTTP/2 connections
if we ask a handler if it implements `http.Hijacker` it returns true, but on invoking `Hijack()` it actually panics and the following short integration test should prove it

#### Server
```go
package main

import (
	"fmt"
	"log"
	"net/http"
	"time"

	"go.opencensus.io/plugin/ochttp"
)

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/tick", tick)
	mux.HandleFunc("/hijack", checkHijack)
	h := &ochttp.Handler{
		Handler: mux,
	}
	if err := http.ListenAndServeTLS(":9944", "cert.pem", "key.pem", h); err != nil {
		log.Fatalf("http.ListenAndServeTLS error: %v", err)
	}
}

func checkHijack(w http.ResponseWriter, r *http.Request) {
	log.Printf("Got connection to checkHijack: %+v\n", r)
	h, ok := w.(http.Hijacker)
	log.Printf("h: %v ok: %v\n", h, ok)
	h.Hijack()
}

func tick(w http.ResponseWriter, r *http.Request) {
	log.Printf("Got connection: %+v\n", r)
	for {
		<-time.After(time.Second)
		fmt.Fprintf(w, "The time right now is: %s\n", time.Now())
		w.(http.Flusher).Flush()
	}
}
```

#### Client
```go
package main

import (
	"crypto/tls"
	"flag"
	"io"
	"log"
	"net/http"
	"os"

	"golang.org/x/net/http2"

	"go.opencensus.io/plugin/ochttp"
)

func main() {
	var route string
	flag.StringVar(&route, "route", "/tick", "the route to check, alternatives are either /tick or /hijack")
	flag.Parse()
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	http2.ConfigureTransport(tr)
	c := &http.Client{
		Transport: &ochttp.Transport{Base: tr},
	}
	c.Transport = tr
	res, err := c.Get("https://localhost:9944" + route)
	if err != nil {
		log.Fatalf("err: %v", err)
	}
	log.Printf("res: %v\n", res)
	io.Copy(os.Stdout, res.Body)
}
```

#### Running it in HTTP/2
1. Streaming should work on visiting /tick -- proving it is an HTTP/2 connection
2. Visiting /hijack should panic the handler -- proving further that we have an HTTP/2 connection whose `ResponseWriter.(http.Hijacker).Hijack()` is a fatal interaction with HTTP/2. This is because  with HTTP/2 we can have multiple requests bidirectionally multiplexed on the same connection while with HTTP/1.X Hijack() takes control of the entire connection and is allowed
3. The HTTP/1.X variant should pass.

The fix is simple, the tests a little more comprehensive but I'll get this in shortly as I chow down on this dinner.

**Comment by t-bast** [comment]
Awesome thanks guys for the quick and thorough investigation.
It's indeed for HTTP/1.x connections, since for HTTP/2 it probably wouldn't make sense to use a websocket because of the full duplex and multiplexing it already offers.

**Comment by odeke-em** [comment]
Thank you for reporting this and for using OpenCensus @t-bast!
I mailed out a fix in #648, we'll merge it in the morning(i.e. a few hours) and you'll be on your way in no time :)

**Comment by t-bast** [comment]
I just had a look at the PR, looks good to me :)
Great work @odeke-em, loving OpenCensus more and more!
